feat(terraform): create detailed documentation (#2323)
FusionAuth/fusionauth-site: site/docs/v1/tech/admin-guide/configuration-management/terraform.adoc

[cols="2,4,4"]
|===
| Strategy | Advantages | Disadvantages

| import | manage existing resources | some resources dont fully support the Terraform lifecycle
| data source  | reference existing resources without having to manage them | requires a external process to manage these resources
| create | create and then manage all resources through Terraform | bigger initial effort
|===

== Terraform Import

Before we go in to Creating, Updating and Removing of resources we first look in to the link:https://developer.hashicorp.com/terraform/language/import[Import] and link:https://developer.hashicorp.com/terraform/language/data-sources[Data Source] functionality. If you have existing resources you want to manage via Terraform you need to import the current configuration to be able to manage it via Terraform.

[NOTE.note]
====
The cleaner way is to create all resources through Terraform instead, but in certain scenarios you will have to use `Import` or `Data Source` and we go through a few examples in this and the next section.
====

> //Additional input for documentation tbd. import cli https://developer.hashicorp.com/terraform/cli/import import config https://developer.hashicorp.com/terraform/language/import https://spacelift.io/blog/importing-exisiting-infrastructure-into-terraform  https://medium.com/swlh/importing-existing-infrastructure-into-terraform-a6ae168ad2bb https://stackoverflow.com/questions/47613926/import-all-resources-defined-in-tf-file https://www.bitslovers.com/terraform-import/)

Author: depending on successfully solving https://github.com/gpsinsight/terraform-provider-fusionauth/issues/214 the document could be extended with fully importing existing config without having to add anything manually.

As we have this issue I can remove the references from the doc for now.

> //Additional input for documentation tbd. import cli https://developer.hashicorp.com/terraform/cli/import import config https://developer.hashicorp.com/terraform/language/import https://spacelift.io/blog/importing-exisiting-infrastructure-into-terraform  https://medium.com/swlh/importing-existing-infrastructure-into-terraform-a6ae168ad2bb https://stackoverflow.com/questions/47613926/import-all-resources-defined-in-tf-file https://www.bitslovers.com/terraform-import/)

Reviewer: Feel free to link to the GH issue directly in the doc if it would be helpful for users.

> //Additional input for documentation tbd. import cli https://developer.hashicorp.com/terraform/cli/import import config https://developer.hashicorp.com/terraform/language/import https://spacelift.io/blog/importing-exisiting-infrastructure-into-terraform  https://medium.com/swlh/importing-existing-infrastructure-into-terraform-a6ae168ad2bb https://stackoverflow.com/questions/47613926/import-all-resources-defined-in-tf-file https://www.bitslovers.com/terraform-import/)

Author: I'll leave it out, as its mentioned at the top that: Since it is a community project, it is not as complete as the client libraries, but it covers the majority of use cases. If you find yourself needing to manage an unsupported resource or configuration, please submit a PR to the project using GitHub.

// The import and generate config command works and I've tested it, but the import configuration has a issue https://github.com/gpsinsight/terraform-provider-fusionauth/issues/214

=== Importing The Default Tenant
Importing the Default tenant is required if you want to manage it through Terraform as it's created during the initial deployment of FusionAuth. In addition, if you are managing FusionAuth application, it will always be part of the Default tenant. As well as there is no need to create a different tenant for creating applications.
//knowledge source https://fusionauth.io/community/forum/topic/1725/what-are-the-pros-and-cons-of-using-the-default-tenant

[NOTE.note]
====
With the usage of Terraform we suggest you to create your own tenant and applications related to your tenant instead of importing.

> ====

Author: This is a bit of a headache to me to be honest, which why it is conflicting.

Importing the existing FusionAuth Tenant and Application is possible but it contradicts somewhat the typical Terraform workflow as its never created by terraform and a non deleteable resource too.

There are some benefits to it especially for a consistent setup, if you want terraform to setup or even further manage certain pieces of the configuration you have to import it, and e.g. can reset what ever a user has changed manually, or specifically ignore where users are allowed to do changes.

The alternative is either creating everything from scratch or using the existing setup as data source and managing the FusionAuth Application and Default Tenant outside of Terraform.

I did in the meantime setup something similar for our own fusionauth environment, and the best experience i had is with creating everything by Terraform including the Tenant. And therefore I personally would suggest not to import to Terraform, but then again it's all depending what you are willing to deal with and what your requirements are.

> ====

Reviewer: Okay. I think if you put this into a paragraph/section and helped the user make the right choice for them (which you kinda did with a table), rather than having two pieces of conflicting advice right next to each other, that would be fine.

* link:https://registry.terraform.io/providers/gpsinsight/fusionauth/latest/docs/data-sources/tenant[Terraform Data Source: Application]
* link:https://registry.terraform.io/providers/gpsinsight/fusionauth/latest/docs/data-sources/application[Terraform Data Source: Application]

[source]
----
data "fusionauth_tenant" "Default" {
  name = "Default"
}

data "fusionauth_application" "FusionAuth" {
  name = "FusionAuth"
}
----

[NOTE.note]
====
If you manage the Default tenant and FusionAuth application outside of Terraform but want specific actions (scripts, API calls, etc.) still integrated and triggered by your Terraform configuration, you could make the use of link:https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax[Provisioners].

Terraform includes the concept of provisioners as a measure of pragmatism and last resort (Since provisioners are non-declarative and potentially unpredictable), knowing that there are always certain behaviors that cannot be directly represented in Terraform's declarative model.
====

> //TODO Provisioner https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax

Author: Sure. I left this as a placeholder, because we could make an example with a other tool managing something in FusionAuth outside of terraform. But then again the remark itself is a placeholder.

== Terraform Create

To create a new resource you can go through the list of resources available to you in the link:https://registry.terraform.io/providers/gpsinsight/fusionauth/latest/docs[FusionAuth Terraform Provider Documentation] and pick the resource you are interested in. Each resource contains information about required and optional arguments.

In this example we create an application called forum in the Default tenant with related roles (admin and user), users (forum-user1 and forum-admin1) and according registrations:

* link:https://registry.terraform.io/providers/gpsinsight/fusionauth/latest/docs/resources/application[Terraform Resource: Application]
* link:https://registry.terraform.io/providers/gpsinsight/fusionauth/latest/docs/resources/application_role[Terraform Resource: Application Role]

> * link:https://registry.terraform.io/providers/gpsinsight/fusionauth/latest/docs/resources/user[Terraform Resource: User]

Author: Its not typical for production environments for sure, but for Test it could be valid. We for example use it to stage the test environment with some default users and additional stuff like fusionauth_registration with prefilled custom data.

With regards to the example I wanted to have something simple which then gets modified very frequently (in the case of user even the first login) to have a easy example on how to use ignore_changes. 

But I can figure out something else instead, as mentioned later the FusionAuth Application data field works similar, it is something you create with terraform but should ignore if you make use of it with your end user application.

> * link:https://registry.terraform.io/providers/gpsinsight/fusionauth/latest/docs/resources/user[Terraform Resource: User]

Reviewer: Makes sense. Let's use application and application.data as something to ignore, please.

* link:https://registry.terraform.io/providers/gpsinsight/fusionauth/latest/docs/resources/registration[Terraform Resource: Registration]

After you are done with your configuration changes, run `terraform plan` to check the planned changes and review potential errors. After that you can go ahead with `terraform apply`.

If you want to know what already has been defined by Terraform but not specified in your `.tf` files you can run `terraform show`.

The configuration can get very large and if you want to show only very specific resources you can list all resources with `terraform state list` and show the resource state with `terraform state show <resource-name>` accordingly.

=== Ignore Lifecycle Changes
//source: ignore lifecycle changes https://itnext.io/how-and-when-to-ignore-lifecycle-changes-in-terraform-ed5bfb46e7ae
Every time you plan new changes you could face the challenge that information has been updated in your FusionAuth installation. You can either align your configuration with the installation or decide to link:https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#ignore_changes[ignore_changes] in your configuration.

> There are two very common examples in FusionAuth resources. One if you manage users through Terraform, where you during creation set password_change_required to true. Or where the user starts changing personal information through FusionAuth or Application.

Reviewer: I think this is a valid reason not to use terraform to manage users :) . I think you could remove this section if we changed the example to not manage users/registrations.

  password_change_required = true
  password                 = "%WLTvrsYELsyPqC^R7FMUNxt##VyDf6XaWk2R7!gS$oL76Ww"
  username_status          = "ACTIVE"
  lifecycle {
    ignore_changes = [
      data,
      image_url,
      password
    ]
  }
}

resource "fusionauth_user" "forum-admin1" {
  email                    = "forum-admin1@email.internal"
  first_name               = "John"
  last_name                = "Doe"
  middle_name              = "William"
  password_change_required = true
  password                 = "@CfosPAVT3&hCzz5c^&#2F5BxNUY$X!@s!7Wx9bd6Yon54e3"
  username_status          = "ACTIVE"
  lifecycle {
    ignore_changes = [
      data,
      image_url,
      password
    ]
  }
}
----

Another example is where your e.g. business application adds user driven data to the FusionAuth Application.

[source]
----
resource "fusionauth_application" "forum" {
  tenant_id = data.fusionauth_tenant.Default.id
  name = "forum"
  lifecycle {
    ignore_changes = [
      data
    ]
  }
}
----

If you are only interested in initially creating and destroying resources you can as well ignore all changes.

[source]
----
lifecycle {
  ignore_changes = all
}
----

== Terraform Remove

> If you want to remove a resource you can comment or delete the lines accordingly. Run `terraform plan` and `terraform apply` once modified which will then destroy the commented or deleted ressource.

Reviewer: ```suggestion
If you want to remove a resource, comment out or delete the lines defining the resource in `main.tf`. Run `terraform plan` to double check. Then run `terraform apply`, which will destroy the resource.
```

> If you run link:https://developer.hashicorp.com/terraform/cli/commands/destroy[terraform destroy] you need to be aware that all Terraform managed resources will be destroyed.

Reviewer: Let's make this a `warning` callout.

=== Prevent Destroy

> If you still want to make sure that certain resources arent destroyed, you can specify the link:https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#prevent_destroy[prevent_destroy] lifecycle meta-argument as a measure of safety against the accidental replacement of objects that may be costly to reproduce. However, it will make certain configuration changes impossible to apply, and will prevent the use of the `terraform destroy` command once such objects are created, and so this option should be used sparingly.

Reviewer: Can we see an example of adding `prevent_destroy` to a (non-default) tenant?

[source]
----
lifecycle {
  prevent_destroy = true
}
----

> A good example would be a productive FusionAuth tenant, as deleting a tenant removes all of the associated applications, groups and users.

Reviewer: ```suggestion
A good example would be an in-use FusionAuth tenant. Deleting a tenant removes all of the associated applications, groups and users.
```

> The same meta-argument you can use for Terraform managed non-deletable resources like Default tenant and FusionAuth application as Terraform would not be able to anyway.

Reviewer: Should we introduce this earlier like when you are importing the fusionauth application and default tenant?

Author: Yes that's a good idea, I will reference this part earlier but I think it's important that the full Prevent Destroy section is in the Terraform Remove section.

Reviewer: Makes sense to repeat it here.
